server/src/main/java/org/elasticsearch/index/shard/IndexWriteLoad.java

```diff
@@ -0,0 +1,217 @@
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+public class IndexWriteLoad implements Writeable, ToXContentFragment {
+    public static final ParseField SHARDS_WRITE_LOAD_FIELD = new ParseField("shards_write_load");
+    public static final ParseField SHARDS_UPTIME_IN_MILLIS = new ParseField("shards_uptime_in_millis");
+    private static final Double UNKNOWN_LOAD = -1.0;
+    private static final long UNKNOWN_UPTIME = -1;
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<IndexWriteLoad, Void> PARSER = new ConstructingObjectParser<>(
+        "index_write_load_parser",
+        false,
+        (args, unused) -> IndexWriteLoad.create((List<Double>) args[0], (List<Long>) args[1])
+    );
+
+    static {
+        PARSER.declareDoubleArray(ConstructingObjectParser.constructorArg(), SHARDS_WRITE_LOAD_FIELD);
+        PARSER.declareLongArray(ConstructingObjectParser.constructorArg(), SHARDS_UPTIME_IN_MILLIS);
+    }
+
+    public static IndexWriteLoad create(List<Double> shardsWriteLoad, List<Long> shardsUptimeInMillis) {
+        if (shardsWriteLoad.size() != shardsUptimeInMillis.size()) {
+            throw new IllegalArgumentException(
+                "The same number of shards write load and shard uptime should be provided, but "
+                    + shardsWriteLoad
+                    + " "
+                    + shardsUptimeInMillis
+                    + " were provided"
+            );
+        }
+
+        if (shardsWriteLoad.isEmpty()) {
+            throw new IllegalArgumentException("At least one shard write load and uptime should be provided, but none was provided");
+        }
+
+        return new IndexWriteLoad(
+            shardsWriteLoad.stream().mapToDouble(shardLoad -> shardLoad).toArray(),
+            shardsUptimeInMillis.stream().mapToLong(shardUptime -> shardUptime).toArray()
+        );
+    }
+
+    @Nullable
+    public static IndexWriteLoad fromStats(IndexMetadata indexMetadata, @Nullable IndicesStatsResponse indicesStatsResponse) {
+        if (indicesStatsResponse == null) {
+            return null;
+        }
+
+        final IndexStats indexStats = indicesStatsResponse.getIndex(indexMetadata.getIndex().getName());
+        if (indexStats == null) {
+            return null;
+        }
+
+        final int numberOfShards = indexMetadata.getNumberOfShards();
+        final var indexWriteLoadBuilder = IndexWriteLoad.builder(numberOfShards);
+        final var indexShards = indexStats.getIndexShards();
+        for (IndexShardStats indexShardsStats : indexShards.values()) {
+            final var shardStats = Arrays.stream(indexShardsStats.getShards())
+                .filter(stats -> stats.getShardRouting().primary())
+                .findFirst()
+                // Fallback to a replica if for some reason we couldn't find the primary stats
```

Comment:
I wonder if we should instead have any fallbacks evaluated at read rather than write time? There we could choose to disregard or fall back to the one with the longer up time etc? Or make decisions based on the wider stream. By amplifying one of the shards results early, we cannot do such things upon read?

Reply:
One issue about fallback during reads is that the write load is "diluted" as we compute it based on the shard uptime and we could potentially query that information long after the latest indexing operation.